examples/shell.rs

```diff
@@ -0,0 +1,236 @@
+
+  if run_shell_flag {
+    run_shell(scope);
+  }
+}
+
+/// Process remaining command line arguments and execute files
+fn run_shell(scope: &mut v8::HandleScope) {
+  use std::io::{self, Write};
+
+  println!("V8 version {} [sample shell]", v8::V8::get_version());
+
+  loop {
+    print!("> ");
+    io::stdout().flush().unwrap();
+
+    let mut buf = String::new();
+    match io::stdin().read_line(&mut buf) {
+      Ok(n) => {
+        if n == 0 {
+          println!();
+          return;
+        }
+
+        execute_string(scope, &buf, "(shell)", true, true);
+      }
+      Err(error) => println!("error: {}", error),
+    }
+  }
+}
+
+/// Process remaining command line arguments and execute files
+fn run_main(
+  scope: &mut v8::HandleScope,
+  args: &[String],
+  run_shell: &mut bool,
+) {
+  let mut skip_next = false;
+
+  // Parse command-line arguments.
+  for (i, arg) in args.iter().enumerate().skip(1) {
+    if skip_next {
+      continue;
+    }
+
+    match &**arg {
+      "--shell" => {
+        // Enables the shell.
+        *run_shell = true;
+      }
+      "-f" => {
+        // Ignore any -f flags for compatibility with the other stand-
+        // alone JavaScript engines.
+      }
+      "-e" => {
+        // Exectute script.
+        let script: &str = &args[i + 1];
+        skip_next = true;
+
+        // TODO: pump event loop (not implemented on rusty_v8?)
+        // while v8::Platform::pump_message_loop(&platform, isolate) {
+        //   // do nothing
+        // }
+        execute_string(scope, script, "unnamed", false, true);
+      }
+      arg => {
+        if arg.starts_with("--") {
+          eprintln!("Warning: unknown flag {}.\nTry --help for options", arg);
+          continue;
+        }
+
+        // Use all other arguments as names of files to load and run.
+        let script = std::fs::read_to_string(arg).expect("failed to read file");
+        execute_string(scope, &script, arg, false, true);
+      }
+    }
+  }
+}
+
+fn execute_string(
+  scope: &mut v8::HandleScope,
+  script: &str,
+  filename: &str,
+  print_result: bool,
+  report_exceptions_flag: bool,
+) {
+  let mut try_catch = v8::TryCatch::new(scope);
+  let mut handle_scope = v8::HandleScope::new(&mut try_catch);
+
+  let script = v8::String::new(&mut handle_scope, script).unwrap();
+  let origin = v8::ScriptOrigin::new(
+    v8::String::new(&mut handle_scope, filename).unwrap().into(),
+    v8::Integer::new(&mut handle_scope, 0),
+    v8::Integer::new(&mut handle_scope, 0),
+    v8::Boolean::new(&mut handle_scope, false),
+    v8::Integer::new(&mut handle_scope, 0),
+    v8::undefined(&mut handle_scope).into(),
+    v8::Boolean::new(&mut handle_scope, false),
+    v8::Boolean::new(&mut handle_scope, false),
+    v8::Boolean::new(&mut handle_scope, false),
+  );
+
+  let script = if let Some(script) =
+    v8::Script::compile(&mut handle_scope, script, Some(&origin))
+  {
+    script
+  } else {
+    drop(handle_scope);
```

Comment:
Just curious, is there a reason to explicitly drop the HandleScope?

Reply:
This is to "trick" Rust borrow checker. `handle_scope` will not be dropped until the program reaches the end of the function.
```
cannot borrow `try_catch` as immutable because it is also borrowed as mutable
immutable borrow occurs hererustc(E0502)
shell.rs(121, 47): mutable borrow occurs here
shell.rs(166, 1): mutable borrow might be used here, when `handle_scope` is dropped and runs the `Drop` code for type `rusty_v8::HandleScope`
```

Reply:
Hm - I wonder if this can be reworked to avoid this drop? Can you use the `scope` argument of the `execute_string()` function instead of creating a new handle scope?

Reply:
Avoided `drop` by using `TryCatch` as a scope.